### DISCUSSION TITLE
1.17.0 Shared behavior issue

### DISCUSSION THREAD

**Discussion by ysmaliak** [pointfreeco/swift-composable-architecture#3536] [General]
### Description

Updating the file storage shared property results in unexpected state changes. The issue was introduced in `1.17.0` and did not occur in `1.16.1`.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Changes to shared property should not affect other state properties.

There are 2 arrays of data: `dummies` and `strings`, which are presented as lists. On `dummies` update, the `strings` should not be affected.

https://github.com/user-attachments/assets/2b9516c7-8de8-402a-82d4-2546acf15081

### Actual behavior

Changes to shared property affect other state properties.

String generation continues, but it's no longer published to view. The `strings` list on the view is reset to an empty state.

https://github.com/user-attachments/assets/8cfedd1c-afc7-4cbd-9798-c0dc5683dc19

### Reproducing project

https://github.com/ysmaliak/ComposableArchitectureSharingTest

There are 4 branches:

-  `main` (uses the `main` branch of the `swift-composable-architecture`, **issue reproducible**)
-  `composable-architecture-1.17.0` (uses version `1.17.0` of the `swift-composable-architecture`, **issue reproducible**)
-  `composable-architecture-1.16.1` (uses version `1.16.1` of the `swift-composable-architecture`, **issue not reproducible**)
-  `pure-swiftui` (uses version `1.1.0` of the `swift-sharing` without `swift-composable-architecture`, **issue not reproducible**)

### The Composable Architecture version information

1.17.0

### Destination operating system

iOS 18.1

### Xcode version information

16.1

### Swift Compiler version information

```shell
swift-driver version: 1.115 Apple Swift version 6.0.2 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)
Target: arm64-apple-macosx15.0
```

**Comment by mbrandonw** [comment]
Hi @ysmaliak, your code is problematic because you are creating a store directly in a view's body: 

```swift
@main
struct ComposableArchitectureSharingTestApp: App {
  var body: some Scene {
    WindowGroup {
      AppView(store: Store(initialState: AppFeature.State(), reducer: { AppFeature() }))
    }
  }
}
```

You should never do this because if the view ever recomputes its body it will recreate the store, thus recreating state from scratch and discarding any in flight effects.

It is better to create the store one single time for the app:

```swift
@main
struct ComposableArchitectureSharingTestApp: App {
  static let store = Store(initialState: AppFeature.State(), reducer: { AppFeature() })
  var body: some Scene {
    WindowGroup {
      AppView(store: Self.store)
    }
  }
}
```

The reason this behavior may have slightly changed in 1.16 vs 1.17 is that `@Shared` may be tracking state access more often as we fixed a few bugs in the process of extracting it out to its own library. So while this may seem like a breaking change, it was really a bug fix, and at the end of the date you should never create stores in views.

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by ysmaliak** [reply]
Didn't realize that. Thanks!

**Reply by holzerryan** [reply]
This also solved an issue for me.  Thank you!
